sg14-error-codes.html

```diff
@@ -604,8 +643,7 @@
     form of the <code>std::type_info</code> singletons which are used by <code>dynamic_cast</code> and also by <code>catch</code>.
 </p>
 <p>
-    Prior to the SG14 telecon, Niall Douglas raised the point that singletons do not play well with DLLs (a.k.a. shared objects,
-    a.k.a. dylibs). On some platforms, there are common programming idioms which can cause a C++ library's "singletons" to become
+    Prior to the SG14 telecon, Niall Douglas raised the point that singletons do not play well with header only libraries. On some platforms, there are common programming idioms which can cause a C++ library's "singletons" to become
```

Comment:
I see you've changed "DLLs" to "header-only libraries" throughout. Is there a problem with header-only libraries when used in programs that are 100% statically linked?

Or do these problems manifest only when the header-only library is used in a DLL? because if so, I'd strongly prefer to continue blaming DLLs. AFAIK, plenty of C++ programmers use header-only libraries without problems, whereas I have never known anyone to use DLLs painlessly.

Reply:
The only showstopper (as in, cannot be worked around) problems are with shared objects on ELF. All other shared library implementations can be beaten into correctness.

Header only libraries, on the other hand, are profoundly broken for using with singletons on all platforms with varying extents of how bad it is. Thus I think you should concentrate on that, because that's a problem with the C++ standard, whereas shared libraries are out of scope for the C++ standard.

BTW Boost next version is redirecting Boost.System to use the std system_error precisely to fix the long standing problems with header only usage.

Reply:
Isn't this something that the C++17 [`inline`](http://en.cppreference.com/w/cpp/language/inline) variables is supposed to help with? For something going into future standard (rather than a defect report), this is available, even if the platform support isn't there yet. It does punt the problem _back_ to shared-library linkage issues, but that's a problem that can be solved at the platform level.

Reply:
At the link level, inline variables are linked exactly the same as inline functions. With ELF, via weak symboling ("last linked gets chosen"), with PE, via selectany ("random choice based on a hash table"). It doesn't do anything at all for the unique-singleton-in-header-only-libraries problem. For that, they need to fix the linker, and so long as they retain using the C linker, that's very hard to achieve.

I've long advocated for the C++ standard standardising its own binary linker. Gaby is heading there eventually with what he has planned after Modules, and I'll be proposing the embedded graph database to drive it next year. We might get there by 2030 or so.